pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1416 @@
+//! A synchronization logic for the [`RelishStorage`] and the state to ensure the correct synchronizations.
+//!
+//! The synchronization does not aim to be immediate, instead
+//! doing all the job in a separate thread asynchronously, attempting to fully replicate the
+//! pageserver timeline workdir data on the remote storage.
+//!
+//! [`SYNC_QUEUE`] is a priority queue to hold [`SyncTask`] for image upload/download.
+//! The queue gets emptied by a single thread with the loop, that polls the tasks one by one.
+//!
+//! During the loop startup, an initial loop state is constructed from all remote storage entries.
+//! It's enough to poll the remote state once on startup only, due to agreement that the pageserver has
+//! an exclusive write access to the relish storage: new files appear in the storage only after the same
+//! pageserver writes them.
+//!
+//! The list construction is currently the only place where the storage sync can return an [`Err`] to the user.
+//! New upload tasks are accepted via [`schedule_timeline_upload`] function disregarding of the corresponding loop startup,
+//! it's up to the caller to avoid uploading of the new relishes, if that caller did not enable the loop.
+//! After the initial state is loaded into memory and the loop starts, any further [`Err`] results do not stop the loop, but rather
+//! reschedules the same task, with possibly less files to sync in it.
+//!
+//! The synchronization unit is an image: a set of layer files (or relishes) and a special metadata file.
+//! Both upload and download tasks consider image in a similar way ([`LocalTimeline`] and [`RemoteTimeline`]):
+//! * a set of relishes (both upload and download tasks store the files as local pageserver paths, ergo [`PathBuf`] is used).
+//! * a set of ids to distinguish the images ([`ZTenantId`] and [`ZTimelineId`])
+//! * `disk_consistent_lsn` which indicates the last [`Lsn`] applicable to the data stored in this image.
+//!
+//! The same relish has identical layer paths in both structs, since both represent the relish path in pageserver's workdir.
+//! This way, the sync can compare remote and local images seamlessly, downloading/uploading missing files if needed.
+//!
+//! `disk_consistent_lsn` and the list of files written gets stored in the metadata file of every image after a successful image checkpoint and
```

Comment:
> and the list of files written gets stored in the metadata file

I can't find this logic anywhere in the code.

Reply:
It's just horrible passive voice usage from me.
What I've meant here is that there's a metadata file with `disk_consistent_lsn`, there's a list of files that were dumped on disk locally, we need both for uploads, both are crated after checkpointing.